[Mangling] Adjust subscript mangling to not include the string "subscript"

ahoppen opened the pull request:
## Problem
The mangled names of subscripts currently contain the string "subscript" which is inconsistent with the representation of subscripts by special names in the AST ([SR-5506](https://bugs.swift.org/browse/SR-5506)). Furthermore, this representation may result in mangled name clashes between subscripts and properties named "subscript" ([SR-5568](https://bugs.swift.org/browse/SR-5568)).

## Implementation 
The approach of this PR is twofold:

1. For the mangling of subscripts themselves simply omit the decl name "subscript" since it is already implied by the node's kind.
2. Mangled accessors currently have the accessed declaration's context, name and type as children. Make accessors have a single child that is either a variable or a subscript node. By 1. subscript nodes then don't contain any name.

## Examples


- Mangling of subscript itself: 
```
old: _T04main3FooC9subscriptAA6OutputVAA5InputVci
new: _T04main3FooC          AA6OutputVAA5InputVci
```
- Mangling of computed property getter:
```
old: _T04main3FooC5myVarAA6MyTypeV fg
new: _T04main3FooC5myVarAA6MyTypeVvfg
```
- Mangling of subscript getter: 
```
old: _T04main3FooC9subscriptAA6OutputVAA5InputVc fg
new: _T04main3FooC          AA6OutputVAA5InputVcifg
```

## Open questions
- Is `OldRemangler` used anywhere? If yes, how can I test it?

ahoppen committed:
[Mangling] Make appendAccessorEntity take an AbstractStorageDecl

ahoppen committed:
[IDE] Change lookup in TypeReconstruction to be by DeclBaseName

ahoppen committed:
[TypeReconstruction] Retrieve subscript setter decl for setter node

ahoppen commented:
@swift-ci Please smoke test

jrose-apple commented:
Thanks for tackling this, Alex! I'll leave most of the review to Erik, but please add tests for:
- private subscripts in non-private types
- non-private subscripts with `private(set)`
- types nested within a subscript accessor
- subscripts with labeled indexes

eeckstein commented on docs/ABI/Mangling.rst:
```diff
@@ -176,9 +176,9 @@ Entities
+  entity-spec ::= type 'i'                           // subscript ('i'ndex) itself (not the individual accessors)
   entity-spec ::= decl-name type 'v'                 // variable
-  entity-spec ::= decl-name type 'f' ACCESSOR
+  entity-spec ::= context entity-spec 'f' ACCESSOR   // enclosed entity is either a variable or subscript
```
-) "context" seems to be wrong here. It's already included "entity ::= context ..."

-) We could do a small optimization here: omit the 'f' for 'i' and 'v'. Instead add a pseudo-accessor which is used for plain variables/subscripts, e.g. 
```
variable-spec ::= type 'i' ACCESSOR
variable-spec ::= decl-name type 'v' ACCESSOR
entity-spec ::= variable-spec
...
ACCESSOR ::= 'p' 
```
Accessor symbols are much more frequent than plain variable symbols. For example:
```
...ip  // instead of ...i
...ig  // instead of ...ifg
```

ahoppen replied:
1.) Oh yes, I didn't notice that

2.) Sound really good. I'll do that

eeckstein reviewed (commented):
Basically looks good to me!

> Is OldRemangler used anywhere? If yes, how can I test it?

Yes. It's used in the runtime to mangle ObjC class names.
You can add tests in test/Interpreter/SDK/RuntimeObjC.swift

Define classes inside variable/subscript getters/setters and check if the NSStringFromClass gives you the same name as before.
See lines 336 ff

ahoppen committed:
[Mangling] Adjust subscript mangling to not include "subscript"

Change the mangling of accessors to have a variable or subscript node
as their only child node, while subscript nodes no longer contain a decl
name.

ahoppen commented:
1.) I have changed the mangling as suggested so that the examples mentioned above now look like this:

- Mangling of subscript itself: 
```
old: _T04main3FooC9subscriptAA6OutputVAA5InputVci
new: _T04main3FooC          AA6OutputVAA5InputVcip
```
- Mangling of computed property getter:
```
old: _T04main3FooC5myVarAA6MyTypeV fg
new: _T04main3FooC5myVarAA6MyTypeVv g
```
- Mangling of subscript getter: 
```
old: _T04main3FooC9subscriptAA6OutputVAA5InputVc fg
new: _T04main3FooC          AA6OutputVAA5InputVci g
```

2.) I have added tests including those requested by @jrose-apple and some ObjC class name tests in `test/IRGen/objc_class_in_storage_decls.swift`. I couldn't use the approach taken in `test/Interpreter/SDK/RuntimeObjC.swift` since classes declared within storage accessors are not accessible on the outer scope that actually gets executed. I'm not quite sure if the tests are placed correctly there but it was the best place I could find.

3.) I have added a `Mangler::verifyOld` function that verifies names mangled with the old mangling scheme, thereby checking that `OldDemangler` can correctly demangle the name again. It's basically a slightly modified copy of `Mangler::verify`. Could you please look through more carefully it since I'm not quite sure if things like checking if the tree contains `Node::Kind::Suffix` and re-demangling is necessary in the old mangling scheme.

ahoppen commented:
@swift-ci Please smoke test

jrose-apple commented on test/IRGen/objc_class_in_storage_decls.swift:
```diff
@@ -0,0 +1,49 @@
+struct WrapperStruct {
+  subscript() -> Void {
+    unsafeAddress {
+      // CHECK: @"OBJC_CLASS_$__TtCFV27objc_class_in_storage_decls13WrapperStructlu9subscriptFT_T_L_31ClassInSubscriptUnsafeAddressor"
```
New tests look good! *Technically* the thing we care about is the string in the class metadata…

```
@5 = private unnamed_addr constant [79 x i8] c"_TtCFV4main13WrapperStructlu9subscriptFT_T_L_31ClassInSubscriptUnsafeAddressor\00"
```

…but this is probably close enough. Thanks, Alex!

(letting Erik handle the actual review)

slavapestov commented:
@ahoppen Can you fix the merge conflict? I think the PR is good to go after that.

ahoppen commented:
@slavapestov I have resolved the merge conflicts and incorporated @jrose-apple advice to check for the string in the class's metadata.

@swift-ci Please smoke test

ahoppen commented:
I just added `REQUIRES: objc_interop` to my new test case `objc_class_in_storage_decls.swift` because it failed under Linux and it was present in all the other ObjC-related tests. I don't know if I implicitly encoded any assumptions about the architecture in the test case at the moment, so maybe someone could take a closer look at that as well.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
Build failure seems to have been unrelated

@swift-ci Please smoke test

ahoppen commented:
I'm getting "undefined symbol" errors for libdispatch on Linux due to my mangling changes. I'm not quite sure if this may be an issue with libdispatch being build incrementally ([ContinuousIntegration.md](https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md) seems to be missing information about this), so I'm just trying a clean build.

@swift-ci Please clean test Linux platform

ahoppen commented:
OK, seems like Linux needs a clean build so that libdispatch gets built with the new mangling scheme. Is that usual or would some version flag need to be incremented because of it?

jrose-apple commented:
@eeckstein, do you know the answer to Alex's question? I assume we just aren't going to worry about this because we aren't ABI-stable yet and we don't care about archived names of local types.

eeckstein commented:
> Is that usual or would some version flag need to be incremented because of it?

This is a known problem (rdar://problem/31097901)

jrose-apple commented:
Sorry, not about the particular issue with libdispatch – a "please clean test" can handle that. Is there anything else tracking the mangling version that Alex would need to update?

ahoppen commented:
I just did a rebase to resolve merge conflicts. If it is OK to work around the libdispatch issue by running "please clean test", I think this PR is good to be merged.

@swift-ci Please clean test

jrose-apple commented:
That's fine. Okay, let's call this ready, then; feel free to merge (with "Create a Merge Commit") once you get the last two tests squashed!

ahoppen commented:
@swift-ci Please clean test

ahoppen commented:
@swift-ci Please clean test

ahoppen commented:
@swift-ci Please clean test

ahoppen commented:
@swift-ci Please clean test
